src/lib/operators/join_helper/join_output_writing.cpp

```diff
@@ -0,0 +1,308 @@
+        // pos_list will contain only NULL_ROW_IDs anyway, so it doesn't matter which Table the ReferenceSegment that
+        // we output is referencing. HACK, but works fine: we create a dummy table and let the ReferenceSegment ref
+        // it.
+        if (!dummy_table) {
+          dummy_table = Table::create_dummy_table(input_table->column_definitions());
+        }
+        output_segments.push_back(std::make_shared<ReferenceSegment>(dummy_table, column_id, pos_list));
+        continue;
+      }
+
+      // Fetch the PosLists for the currently processed column. We can later use this list to avoid extracting
+      // PosList again from an AbstractSegment and it provides us with a key for `output_pos_list_cache`.
+      const auto& input_table_pos_lists = input_pos_list_ptrs_sptrs_by_segments[column_id];
+
+      auto iter = output_pos_list_cache.find(input_table_pos_lists);
+      if (iter == output_pos_list_cache.end()) {
+        // Get the row ids that are referenced
+        auto new_pos_list = std::make_shared<RowIDPosList>();
+        new_pos_list->reserve(pos_list->size());
+
+        auto common_chunk_id = std::optional<ChunkID>{};
+        for (const auto& row : *pos_list) {
+          if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
+            new_pos_list->push_back(row);
+            common_chunk_id = INVALID_CHUNK_ID;
+          } else {
+            const auto& referenced_pos_list = *(*input_table_pos_lists)[row.chunk_id];
+            new_pos_list->push_back(referenced_pos_list[row.chunk_offset]);
+
+            // Check if the current row matches the ChunkIDs that we have seen in previous rows
+            const auto referenced_chunk_id = referenced_pos_list[row.chunk_offset].chunk_id;
+            if (!common_chunk_id) {
+              common_chunk_id = referenced_chunk_id;
+            } else if (*common_chunk_id != referenced_chunk_id) {
+              common_chunk_id = INVALID_CHUNK_ID;
+            }
+          }
+        }
+        if (common_chunk_id && *common_chunk_id != INVALID_CHUNK_ID) {
+          // Track the occuring chunk ids and set the single chunk guarantee if possible. Generally, this is the case
+          // if both of the following are true: (1) The probe side input already had this guarantee and (2) no radix
+          // partitioning was used. If multiple small PosLists were merged (see MIN_SIZE in join_hash.cpp), this
+          // guarantee cannot be given.
+          new_pos_list->guarantee_single_chunk();
+        }
+
+        iter = output_pos_list_cache.emplace(input_table_pos_lists, new_pos_list).first;
+      }
+
+      auto reference_segment =
+          std::static_pointer_cast<const ReferenceSegment>(input_table->get_chunk(ChunkID{0})->get_segment(column_id));
+      output_segments.push_back(std::make_shared<ReferenceSegment>(
+          reference_segment->referenced_table(), reference_segment->referenced_column_id(), iter->second));
+    } else {
```

Comment:
For this case (changed one above, curious why tidy wasn't complaining) I already though it, but I wouldn't use it here.
For guard clauses, it makes a lot of sense. Here, we have two main paths in this function, the if-else makes that very clear.

Do you think it would be more readable without the else?

Reply:
Just thought about indentation levels and so on. If you don't collapse the if branch, I think you are also lost to which if the else actually belongs, so no benefit anymore. But nothing I would fight for. Mybe add a comment where we are now and what is going on here

Reply:
Collapsing sounds like you are using an IDE. I won't talk to you.

Reply:
Don't use the evil three letters 😨 
A text editor does this, as well. As I said, nothing too hinged on. We don't use this weird novel functional and object-oriented stuff in an imperative language anyway, do we? :D
Back to topic, get your point. Do as you like here I guess 🤷‍♀️